searx/settings_defaults.py

```diff
@@ -0,0 +1,172 @@
+# SPDX-License-Identifier: AGPL-3.0-or-later
+# lint: pylint
+# pylint: disable=missing-function-docstring, missing-module-docstring
+
+import os
+import sys
+import logging
+from os.path import dirname, abspath
+
+searx_dir = abspath(dirname(__file__))
+
+logger = logging.getLogger('searx')
+OUTPUT_FORMATS = ['html', 'csv', 'json', 'rss']
+
+
+def default(value):
+    def apply(existing, _path):
+        return value if not existing else existing
```

Comment:
> Why does the caller of the closure calls the closure with unneeded arguments?

To be able to log a warning:
https://github.com/searxng/searxng/blob/cc05240031cc3ab3cfb67173376d96c39e55252e/searx/settings_defaults.py#L52

> Does it make sense to catch all unneeded arguments by def inner(existing, *_args, **_kwargs)

The returned closures are called by this line, and only by this line:
https://github.com/searxng/searxng/blob/cc05240031cc3ab3cfb67173376d96c39e55252e/searx/settings_defaults.py#L74 

----

The idea of `apply_schema`:
* for each `key`, `value` of the `schema` do `settings.setdefault(key, value)` with recursive calls if `value` is a `dict`.
* since there is a recursive call, `settings` here might actually be `settings['ui']['theme_args']` (for example).
* but there are two exceptions:
   * if `value` is `callable`or `tuple` (both won't be found in `settings`)
   * if `value` is a `tuple` the code expects that all items of tuple are `callable`
   * for each `callable` named `f` do `settings[key]=f(settings[key], path)`.

----

Perhaps a class like with a constructor like this:
`SettingsValue(type_list=[str], environ_name='SEARX_BIND_ADDRESS', default='127.0.0.1')`
is better (all arguments are optional).

Reply:
> The returned closures are called by this line, and only by this line:

Ah, this was the point I missed / thanks. Another point I missed: `settings` is the dictionary from `searx.settings_loader.load_settings()` .. I think it would be more _readable_ when we have a class which implements all the _settings_ stuff.

> Perhaps a class like with a constructor like this:

See my example, not a `SettingsValue` class, I mean a `Settings` which implements the whole setting stuff (this from `settings_loader` and this from `settings_defaults`.

With the hints you gave above I can try to implement such a class in a separated branch (the branch will start on top of your's). But I can't predict when I have time for.

Reply:
FYI: started to implement a class `Settings` .. it s far from finish but might be illustrate what I mean, see my [mod-default-settings](https://github.com/return42/searxng/commits/mod-default-settings) branch.

Reply:
Few comments on the spot / idea:
* What do you think of a module `searx.settings` so the loader part is in different file ? Purpose: avoid a huge file doing everything.
* `settings = Settings(...)`  then `settings['ui']['static_path'] should work as before, otherwise it will requires too much changes everywhere.
* but  for example `get_themes`, `get_static_files` and `get_result_templates` (from searx.webutils) can move into this class in a way another.

Reply:
> What do you think  ...

Yes, this is what I also thought, at least _more or less_  because I haven't (can't) yet thought all to its end.

> IMO, If it includes the change from this PR, it should be another PR 

Yes, lets first finish this PR .. afterwards we can discuss how to assemble all or only a part in a new settings.py file (module).